### PR TITLE
core: fix terminated status check in TransportListener

### PR DESCRIPTION
**Pull request by creamsoup**
when we queue and execute the runnable in transport terminated, it may not be drained immediately if there's a already drain is running on the other thread. 
Adding a boolean to check shutdown is called before terminated.

### CODE REVIEW

**Comment by creamsoup** [comment]
internally tracked bug b/136496833

File: core/src/main/java/io/grpc/internal/InternalSubchannel.java [hunk 1]
```diff
@@ -573,7 +575,7 @@ public void run() {
           }
         }
       });
-      Preconditions.checkState(activeTransport != transport,
+      Preconditions.checkState(shutdownInitiated || activeTransport != transport,
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/InternalSubchannel.java]: Just check `shutdownInitiated`, and you can do it at the beginning of the method.

Comment (creamsoup) [reply core/src/main/java/io/grpc/internal/InternalSubchannel.java]: thanks done. had to fix few tests PTAL

File: core/src/test/java/io/grpc/internal/InternalSubchannelTest.java [hunk 2]
```diff
@@ -1145,6 +1146,8 @@ public void channelzMembership() throws Exception {
 
     MockClientTransportInfo t0 = transports.poll();
     assertTrue(channelz.containsClientSocket(t0.transport.getLogId()));
+    t0.listener.transportShutdown(Status.RESOURCE_EXHAUSTED);
+    fakeClock.forwardNanos(10);
```
Comment (ejona86) [inline core/src/test/java/io/grpc/internal/InternalSubchannelTest.java]: Why is advancing the time necessary?

File: core/src/test/java/io/grpc/internal/ManagedChannelImplTest.java [hunk 3]
```diff
@@ -521,8 +521,10 @@ public void channelzMembership_subchannel() throws Exception {
     MockClientTransportInfo transportInfo = transports.poll();
     assertNotNull(transportInfo);
     assertTrue(channelz.containsClientSocket(transportInfo.transport.getLogId()));
+    transportInfo.listener.transportReady();
```
Comment (ejona86) [inline core/src/test/java/io/grpc/internal/ManagedChannelImplTest.java]: Adding this to this test looks fine/natural (so no need for a change). I will just note that it shouldn't be _necessary_ for the transport to become ready before being shutdown. That happens during negotiation failure, for example.

File: core/src/test/java/io/grpc/internal/InternalSubchannelTest.java (hunk 2 above)
Comment (creamsoup) [reply core/src/test/java/io/grpc/internal/InternalSubchannelTest.java]: shutdown triggers connection backoff which won't be executed if not advancing the time. this fails the check in After block (checking any pending tasks)

File: core/src/test/java/io/grpc/internal/ManagedChannelImplTest.java (hunk 3 above)
Comment (creamsoup) [reply core/src/test/java/io/grpc/internal/ManagedChannelImplTest.java]: this is also related to reconnect, the shutdown triggers backoff and reconnect so the subchannel is not removed from channelz which makes the test fail.

File: core/src/test/java/io/grpc/internal/InternalSubchannelTest.java (hunk 2 above)
Comment (ejona86) [reply core/src/test/java/io/grpc/internal/InternalSubchannelTest.java]: Go ahead and issue the ready, so the shutdown doesn't trigger a backoff. That is the more "normal" flow.

Comment (creamsoup) [reply core/src/test/java/io/grpc/internal/InternalSubchannelTest.java]: okay. done.

**Pull request merged**